File: calendar_integration.py
# calendar_integration.py
import datetime
import os
from google.oauth2 import service_account
from googleapiclient.discovery import build
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

SCOPES = ['https://www.googleapis.com/auth/calendar']


def get_calendar_service():
    """
    Returns an authorized Google Calendar service instance.
    """
    creds_file = os.getenv("GOOGLE_CALENDAR_CREDENTIALS_FILE")
    if not creds_file:
        raise EnvironmentError("GOOGLE_CALENDAR_CREDENTIALS_FILE environment variable is not set.")
    creds = service_account.Credentials.from_service_account_file(creds_file, scopes=SCOPES)
    service = build('calendar', 'v3', credentials=creds)
    return service


def add_event(event_info: dict):
    """
    Creates a calendar event using the provided event_info dictionary which contains
    'event' (summary) and 'date' (in MM/DD/YYYY format).
    Returns the created event, or None on failure.  Includes more robust error handling.
    """
    service = get_calendar_service()

    try:
        event_date = datetime.datetime.strptime(event_info["date"], "%m/%d/%Y").date()
    except ValueError:
        print(
            f"Date format not recognized for '{event_info['date']}' in event '{event_info['event']}'. Skipping."
        )
        return None

    event = {
        'summary': event_info["event"],
        'start': {
            'date': event_date.isoformat(),
        },
        'end': {
            'date': (event_date + datetime.timedelta(days=1)).isoformat(),
        },
    }

    try:
        calendar_id = os.getenv("CALENDAR_ID")
        if not calendar_id:
            raise EnvironmentError("CALENDAR_ID environment variable is not set.")
        created_event = service.events().insert(calendarId=calendar_id, body=event).execute()
        return created_event
    except Exception as e:
        print(f"Failed to add event '{event_info['event']}' on {event_info['date']}: {e}")
        return None


def add_events(events: list) -> list:
    """
    Given a list of event dictionaries, adds each event to the calendar.
    Returns a list of successfully added events.
    """
    added_events = []
    for event_info in events:
        created = add_event(event_info)
        if created:
            added_events.append(created)
    return added_events

File: google_calendar/calendar_integration.js
import { google } from 'googleapis';
import { auth } from 'google-auth-library';
import dotenv from 'dotenv';
import fs from 'fs';

// Load environment variables
dotenv.config();

const SCOPES = ['https://www.googleapis.com/auth/calendar'];

// ===============================
// calendar_integration.js (Translated)
// ===============================

/**
 * Returns an authorized Google Calendar service instance.
 * @returns {Promise<google.calendar_v3.Calendar>} - A Google Calendar service instance.
 */
async function getCalendarService() {
    const credsFile = process.env.GOOGLE_CALENDAR_CREDENTIALS_FILE;
    if (!credsFile) {
        throw new Error("GOOGLE_CALENDAR_CREDENTIALS_FILE environment variable is not set.");
    }

    try {
        const credentials = JSON.parse(fs.readFileSync(credsFile, 'utf8'));
        const client = auth.fromJSON(credentials);
        client.scopes = SCOPES;

        const service = google.calendar({ version: 'v3', auth: client });
        return service;
    } catch (error) {
        console.error('Error loading credentials:', error);
        throw error;
    }
}

/**
 * Creates a calendar event using the provided event_info dictionary.
 * @param {{event: string, date: string}} eventInfo - Object containing event details.
 * @returns {Promise<any | null>} - A promise that resolves to the created event, or null on failure.
 */
async function addEvent(eventInfo) {
    const service = await getCalendarService();

    try {
        // Parse the date string (MM/DD/YYYY format)
        const [month, day, year] = eventInfo.date.split('/').map(Number);
        const eventDate = new Date(year, month - 1, day);
        
        if (isNaN(eventDate.getTime())) {
            console.error(`Date format not recognized for '${eventInfo.date}' in event '${eventInfo.event}'. Skipping.`);
            return null;
        }

        const event = {
            summary: eventInfo.event,
            start: {
                date: eventDate.toISOString().split('T')[0],
            },
            end: {
                date: new Date(eventDate.setDate(eventDate.getDate() + 1)).toISOString().split('T')[0],
            },
        };

        const calendarId = process.env.CALENDAR_ID;
        if (!calendarId) {
            throw new Error("CALENDAR_ID environment variable is not set.");
        }

        const createdEvent = await service.events.insert({
            calendarId: calendarId,
            requestBody: event,
        });
        return createdEvent.data;
    } catch (error) {
        console.error(`Failed to add event '${eventInfo.event}' on ${eventInfo.date}: ${error}`);
        return null;
    }
}

/**
 * Adds a list of events to the calendar.
 * @param {Array<{event: string, date: string}>} events - An array of event objects.
 * @returns {Promise<Array<any>>} - A promise that resolves to an array of successfully added events.
 */
async function addEvents(events) {
    const addedEvents = [];
    for (const eventInfo of events) {
        const created = await addEvent(eventInfo);
        if (created) {
            addedEvents.push(created);
        }
    }
    return addedEvents;
}

export { addEvents, addEvent, getCalendarService };